Library/Homebrew/livecheck/strategy/sparkle.rb

```diff
@@ -86,6 +86,16 @@ def self.item_from_content(content)
 
             next if (os = enclosure&.attr("os")) && os != "osx"
 
+            if OS.mac? && (minimum_system_version = (item > "minimumSystemVersion").first&.text&.strip)
+              macos_minimum_system_version = begin
+                MacOS::Version.new(minimum_system_version).strip_patch
+              rescue MacOSVersionError
+                nil
+              end
+
+              next if MacOS.version < macos_minimum_system_version
+            end
+
```

Comment:
1. sparkle strategy applies to casks only and I wouldn't worry about them on Linux
2. I don't think we should be comparing things to the version of macOS the user is running (MacOS.version)
3. Unlike in Homebrew/installI, we don't have `MACOS_NEWEST_UNSUPPORTED` in Homebrew/brew, so currently we have to use a hardcoded value 
4.  I think the same result can be achieved with:
```suggestion
            min_sys_ver = (item > "minimumSystemVersion").first
            minimum_system_version = min_sys_ver.present? ? min_sys_ver.text : "0.0.0"
            required_macos = MacOS::Version.new(minimum_system_version)
            newest_unsupported = MacOS::Version.new("12.0")
            next if required_macos >= newest_unsupported
```

Reply:
The main issues I have with alternative are:
- Hardcoding value means we need to update this annually, which adds extra maintenance overhead.
- It is possible that a cask uses a non-standard tag. For example, `omnipresence` uses its own special `omniappcast:minimumSystemVersion`. Since we ignore XML namespaces, the value is passed into `MacOS::Version.new` and causes unwanted exception.

Still unsure whether better or worse to look at system OS version.
As previously mentioned, there are pros and cons to it.

---

Taking a situation like:
```
  if MacOS.version <= :catalina
    version "1.0.1"
  else
    version "2.0.1"
  end

  livecheck do
    url "..."
    strategy :sparkle
  end
```

Running `brew livecheck` or `brew audit --online` on Catalina may show a "new" version.

This can lead to a user trying `brew bump-cask-pr ...` and potentially requesting update to wrong version.

---

On other hand, let's say we have a situation like
```
  version "2.0.1"

  livecheck do
    url "..."
    strategy :sparkle
  end

  depends_on macos: ">= :catalina"
```

Let's say upstream releases "3.0.0" with minimum macOS of Big Sur (11.0).

Ignoring local OS allows for a consistent output whether on Catalina or Big Sur:
- `2.0.1 -> 3.0.0`

Some sparkle feeds also have only the newest release listed, which means that analyzing OS can cause livecheck to show an error on Catalina.

Reply:
I amended my original suggestion to take care of the leading 'v' character as well as a few other things. In my tests it seems to work well. I agree with that hardcoded values are not good, but currently we don't have apparatus to get the "newest unsupported" macOS version, so perhaps this is something we could implement. For example, we could define `:newest_unsupported` symbol in `os/mac/version.rb`. 

Feel free to use my suggestion as a starting point or ignore it. My main point was that we probably shouldn't use `OS.mac?` conditional because casks don't work on Linux. I also think the result of `livecheck` should not depend on the OS version of the user issuing the command -- it might lead to the debugging hell at some point.

Reply:
> I don't think we should be comparing things to the version of macOS the user is running (MacOS.version)

> I also think the result of livecheck should not depend on the OS version of the user issuing the command -- it might lead to the debugging hell at some point.

I agree. `brew livecheck` should return the newest stable version [intended for a stable macOS version] regardless of the execution environment. This PR is about avoiding a version in a Sparkle feed that's only intended for a beta version of macOS, so we only need to know the latest stable macOS version to avoid it.

If this PR didn't exist, I would say we should just switch the `livecheck` block in `bartender` to checking the version headings of the [Bartender 4 release notes page](https://www.macbartender.com/Bartender4/release_notes/), as that lists `4.0.48` as newest (this is also the version in the disk image file on the [Support page](https://www.macbartender.com/Bartender4/support/)). [For what it's worth, I have a stashed `livecheck` block for this.]

I don't imagine this particular Sparkle situation will apply to many casks over time but it probably wouldn't hurt to handle this automatically. Out of curiosity, have you seen this type of issue in any casks using the `Sparkle` strategy other than `bartender`?

---

Regarding this code, using `MacOS::Version.from_symbol(:big_sur)` would be better than `MacOS.version` but, as was mentioned, this would have to be manually updated when a new major macOS version is officially released.

> We could look at the newest OS value already in `os/mac/version.rb` or some other file, though it does show `:monterey` now.

We can't rely on the first symbol in `MacOS::Version::SYMBOLS` being either the current developer beta or the current stable version, as this will vary depending on if we're inside or outside of the window from WWDC to the next major macOS release.

We already manually update `MacOS::Version` for new beta macOS releases, so it may make sense to modify it to distinguish between stable/unstable versions. This would probably stand a better chance of being updated compared to a symbol in the `Sparkle` strategy.

This setup could look like this:

```ruby
UNSTABLE_SYMBOLS = {
  monterey: "12",
}.freeze

STABLE_SYMBOLS = {
  big_sur:     "11",
  catalina:    "10.15",
  mojave:      "10.14",
  high_sierra: "10.13",
  sierra:      "10.12",
  el_capitan:  "10.11",
  yosemite:    "10.10",
}.freeze

SYMBOLS = UNSTABLE_SYMBOLS.merge(STABLE_SYMBOLS).freeze
```

When the latest unstable version is officially released, we simply move the symbol from the unstable hash to the top of the stable hash. So long as the symbols are kept in order from newest to oldest, we could add a simple `#latest_stable` method to `OS::Mac` that uses `STABLE_SYMBOLS`:

```ruby
sig { returns(Version) }
def latest_stable
  first_symbol = Version::STABLE_SYMBOLS&.first&.first
  first_symbol.present? ? Version::from_symbol(first_symbol) : nil
end
```

This would allow us to use `next if MacOS.latest_stable < macos_minimum_system_version` in the code here.

What do we think about this approach? If it's amenable, I'll open a PR for the changes and then we can come back to update this.

Reply:
I like both suggestions -- `latest_stable` method and STABLE/UNSTABLE symbols. The only part I'm a bit skeptical about is using order-dependent operations (`first.first`) on a Hash. That might be OK, though -- idk.

Reply:
> The only part I'm a bit skeptical about is using order-dependent operations (`first.first`) on a Hash. That might be OK, though -- idk.

It works correctly with the aforementioned setup but it's easy to break, so I understand the concern.

We can also just use a hardcoded value, following suit with the `OS::Mac#outdated_release?` and `OS::Mac#prerelease?` methods:

```ruby
sig { returns(Version) }
def latest_stable
  # TODO: bump version when new macOS is released
  Version::from_symbol(:big_sur)
end
```

This wouldn't require any changes to `SYMBOLS` and this method would be next to the other methods that need to be updated when a new macOS release happens.

Reply:
Yet another alternative, we could simply modify `OS::Mac`/`OS::Mac::Version` to make the `#prerelease?` method work for a version created from a string using `OS::Mac::Version#new` (not just `OS::Mac.prerelease?`, which uses the execution environment's macOS version). This would allow us to do `next if macos_minimum_system_version.prerelease?` instead.

I opened a PR for this (#11581), as it's ever-so-slightly less work than `OS::Mac#latest_stable` would be. `#prerelease?` compares the existing `::Version` to a string, whereas `#latest_stable` has to create a new `::Version` for us to do our own comparison. I think recycling `#prerelease?` is the way to go for now but let's see how that PR fares.